frontend/README.md

```diff
@@ -0,0 +1,83 @@
+# Stable Diffusion Web UI
+
+Demo at https://peaceful-otter-7a427f.netlify.app/ (not connected to back end)
+
+much of this readme is just notes for myself during dev work
+
+## Test and Build
+
+from `frontend/`:
+
+-   `yarn dev` runs `tsc-watch`, which runs `vite build` on successful `tsc` transpilation
+
+from `.`:
+
+-   `python backend/server.py` serves both frontend and backend at http://localhost:9090
+
+## API
+
+`backend/server.py` serves the UI and provides a [socket.io](https://github.com/socketio/socket.io) API via [flask-socketio](https://github.com/miguelgrinberg/flask-socketio).
+
+### Server Listeners
+
+The server listens for these socket.io events:
+
+`cancel`
+
+-   Cancels in-progress image generation
+-   Returns ack only
+
+`generateImage`
+
+-   Accepts object of image parameters
+-   Generates an image
+-   Returns ack only (image generation function sends progress and result via separate events)
+
+`deleteImage`
+
+-   Accepts file path to image
+-   Deletes image
+-   Returns ack only
+
+`deleteAllImages` WIP
+
+-   Deletes all images in `outputs/`
+-   Returns ack only
+
+`requestAllImages`
+
+-   Returns array of all images in `outputs/`
+
+`requestCapabilities` WIP
+
+-   Returns capabilities of server (torch device, GFPGAN and ESRGAN availability, ???)
+
+`sendImage` WIP
+
+-   Accepts a File and attributes
+-   Saves image
+-   Used to save init images which are not generated images
+
+### Server Emitters
+
+`progress`
+
+-   Emitted during each step in generation
+-   Sends a number from 0 to 1 representing percentage of steps completed
+
+`result` WIP
+
+-   Emitted when an image generation has completed
+-   Sends a object:
+
+```
+{
+    url: relative_file_path,
+    metadata: image_metadata_object
+}
+```
+
+## TODO
+
+-   Search repo for "TODO"
+-   My one gripe with Chakra: no way to disable all animations right now and drop the dependence on `framer-motion`. I would prefer to save the ~30kb on bundle and have zero animations. This is on the Chakra roadmap. See https://github.com/chakra-ui/chakra-ui/pull/6368 for last discussion on this. Need to check in on this issue periodically.
```

Comment:
Besides having animations baked-in, Chakra is also built on top of emotion, which is a CSS-in-JS library. 

Because it's computing CSS styles at runtime, and also during re-renders when styles change, I'm afraid it's going to be a mild performance hog no matter what. Most React UI component libraries that come with styles baked-in will be like this though :confounded:  

Not sure if it will be the right choice here, but I've been enjoying the results from this CSS-only UI framework, [daisyUI](https://daisyui.com), in my webui. And it was really easy to [disable the animations](https://github.com/nderscore/solid-diffusion/blob/main/src/index.css#L101-L115) globally.

Reply:
Thanks for the suggestion. I'll look into daisyUI, CSS-only sounds great. Do you have any other suggestions for UI frameworks that are lightweight and have a good selection of components? I've only used Chakra because I was familiar with it and it provides a lot of stuff out of the box, certainly not married to it.

Reply:
I spent some time poking around at a few options the other week, and daisyUI (while it comes with a little bit of markup jank to stay pure to it's css-only vision sometimes :confounded: ) provided the most broad selection of components by a long-shot. 

It's also a tailwind plugin, which should feel somewhat similar to Chakra in it's usage of scales except all of the props are css utility classes and [responsive styles](https://chakra-ui.com/getting-started/comparison#responsive-styles-) aren't arrays. For some inspiration about how it could be implemented, I would check out [my repo](https://github.com/nderscore/solid-diffusion/) - it's SolidJS, but since it's all function components, hooks, and JSX, so it should look ~95% similar to React code anyway

I found a few other options for CSS-only frameworks, but a lot of them were either too minimalist, or lacked some UI components I thought would be needed (some that I remember: [PureCSS](https://purecss.io), [PicoCSS](https://picocss.com), [milligram](https://milligram.io)). I guess there's always boostrap too :nauseated_face: 

---

There's also the option of using a headless (style-less) component library in conjunction with any of these CSS frameworks.

[Radix UI](https://www.radix-ui.com) and [Headless UI](https://headlessui.com) both provide for accessibility and state management of some of the most common patterns.

Reply:
Thanks for the suggestions. Though I don't know tailwind, I do know it is very popular and would facilitate contributions to the project. Chakra is pretty popular now it seems and is fairly mature. 

At this point I'm fairly far along in the design, so the time investment to switch to a different UI library will be similar if done now vs later, so I'll stick with Chakra while I sort out the re-rendering issues.

Reply:
Tailwind is a great library but I am personally not for it because it adds a ton of in-line classes that makes the JSX quite obnoxious to read through. Anyone who wants to contribute to the code later will find it really difficult to process the code if they're not familiar with Tailwind already.

The same applies to any other styling library. Which is why I am great fan of using CSS with a preprocessor like SASS. If you are comfortable with CSS, there is nothing that can be better than this for custom styling even if it means you take the little extra time to write the code yourself.

Reply:
> because it adds a ton of in-line classes that makes the JSX quite obnoxious to read through

I was going to mention this, but I wasn't certain that I was remembering correctly. It *really* pollutes the CSS namespace and makes in-browser debugging difficult (surely there's some way to tree-shake it down to just the stuff that's actually in use?)

Reply:
> because it adds a ton of in-line classes that makes the JSX quite obnoxious to read through

As long as it's tucked away into a reusable atomic component, it's not so bad. 

>  (surely there's some way to tree-shake it down to just the stuff that's actually in use?)

This is something tailwind does. Only classes that are referenced in the source files are included in the final CSS.

Using tailwind also isn't required to use daisyUI. It does integrate with it though, since it's implemented as a tailwind plugin (and using it with tailwind lets that tree-shaking of unused classes take place)

> Which is why I am great fan of using CSS with a preprocessor like SASS. If you are comfortable with CSS, there is nothing that can be better than this for custom styling even if it means you take the little extra time to write the code yourself.

Writing everything from scratch means encountering and fixing a lot of bugs along the way that are already solved problems, but I agree that something closer to pure CSS would be more approachable. SASS can also get a bit incoherent in its own way though, especially if you start to utilize the more advanced features (functions, mixins, conditionals, loops, etc). It would definitely avoid the performance concerns related to CSS-in-JS styles though.

No solution will be perfect though. Styling for React apps is an unsolved problem and there countless competing ways to do it, each with their own trade-offs.